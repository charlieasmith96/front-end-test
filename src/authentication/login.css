.login-container {
    margin: 0 auto;
    margin-top: 20%;
    width: 400px;
    border-radius: 10px;
    background-color: #0A3758;
}

.login-title {
    padding-top: 20px;
    text-align: center;
    color: white;
    font-family: Helvetica, Arial;
    font-weight: 20px;
}

.login {
    margin: 0 auto;
    width: 120px;
}

.login-button {
    width: 80px;
    height: 30px;
    margin: 20px 0 20px 20px;
    border-radius: 10px;
    font-size: 16px;
    color: white;
    background-color: #DDA02A;
    border: none;
    transition: 0.3s;
}

.login-button:hover {
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.5);
    transition: 0.3s;
}

.field-label{ 
    color: white;
    text-shadow: none;
    margin-left: 10px;
}

.background { 
    height: 100%;
    width: 100%;
    position: absolute;
    background: url(../static-assets/login-background.png) no-repeat center center fixed; 
    -webkit-background-size: cover;
    -o-background-size: cover;
    background-size: cover;
  }

  .form-field {
      height: 20px;
      border-radius: 5px;
      font-size: 15px;
      width: 126px;

  }