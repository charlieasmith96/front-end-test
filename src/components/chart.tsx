import React, { useState } from 'react';
import { createContainer , VictoryLine, VictoryArea, VictoryStack, VictoryChart, VictoryBrushContainer, VictoryAxis } from 'victory';
import { subtractDate } from '../helpers/date-utils';

export interface ChartProps {
    dataSet1?: BroadcastData
    dataSet2?: BroadcastData 
    startDate: number
    endDate: number
    p2pAndCdnMax?: number 
    cdnMax?: number 
}

const VictoryZoomVoronoiContainer = createContainer("zoom", "voronoi");
const INITIAL_ZOOM_DOMAIN = { x: [subtractDate(4), subtractDate(2)] }

export type BroadcastData = number[][]

export const Chart = (props : ChartProps)  => {
    const [zoomDomain, setZoomDomain] = useState(INITIAL_ZOOM_DOMAIN);

    const handleZoom = (domain: any) => {
      domain.y = undefined;
      setZoomDomain(domain);
    }
    
    return (
        <div>
            <VictoryChart width={1000} height={250} scale={{x: "time"}} 
            containerComponent={
              <VictoryZoomVoronoiContainer 
                //@ts-ignore
                zoomDimension="x"
                allowZoom={false}
                allowPan={false}
                zoomDomain={zoomDomain}
                onZoomDomainChange={(domain : any) => handleZoom(domain)}
                //@ts-ignore
                labels={({datum }) => `Maximum: ${(datum[1])}`}
                voronoiDimension="x"
              />
            }>

            <VictoryStack>
                <VictoryArea interpolation={"natural"} data={props.dataSet1} x={0} y={1} style={{data: {fill: "#C42151"}}} />
                <VictoryArea interpolation={"natural"} data={props.dataSet2} x={0} y={1} style={{data: {fill: "#12A5ED"}}} />
             </VictoryStack>

             <VictoryLine data={[[props.startDate - 1, props.cdnMax], [props.endDate - 1, props.cdnMax]]}
              x={0} y={1} style={{data: {stroke: "red"}}} />

             <VictoryLine data={[[props.startDate - 1, props.p2pAndCdnMax], [props.endDate - 1, props.p2pAndCdnMax]]}
              x={0} y={1} style={{data: {stroke: "green", strokeWidth: 3}}}/>

            </VictoryChart>
            <VictoryChart width={1000} height={100} scale={{x: "time"}} padding={{top: 0, left: 50, right: 50, bottom: 30}}
            style={{background: { fill: "#C9EDD9" }}}
            containerComponent={
              <VictoryBrushContainer responsive={true}
              brushDimension="x"
              //@ts-ignore
              brushDomain={zoomDomain}
              onBrushDomainChange={(domain : any) => handleZoom(domain)}
              />
            }>
            <VictoryAxis>

            </VictoryAxis>
              <VictoryLine data={props.dataSet1} x={0} y={1} interpolation={"natural"} />
              <VictoryLine data={props.dataSet2} x={0} y={1} interpolation={"natural"} />
            </VictoryChart>
        </div>
    )
}