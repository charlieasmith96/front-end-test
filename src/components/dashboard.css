.chart-container {
    width: 80%;
    margin: 0 auto;
}

.right-dropdown {
    right: 20px;
}

.dropdown-container {
    justify-content: space-between;
    display: flex;
}

.axis-label {
    display: block;
    margin: 0;
    width: 10px;
    position: absolute;
    margin-top: 30px;
}


.dashboard-error {
    padding-top: 250px;
    font-size: 30px;
}