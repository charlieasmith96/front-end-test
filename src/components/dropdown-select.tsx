import React from 'react';
import { Form } from 'react-bootstrap';

interface DropdownSelectProps {
    onChangeHandler: (e: any) => void,
    dates: Date[],
    selectedIndex: number,
    className? : string
}

export const DropdownSelect = (props: DropdownSelectProps) => (
  <Form.Control className={props.className} as="select" onChange={(e) => props.onChangeHandler(e)}>
    {props.dates.map((date, index, array) => {
      return (
        <option
          selected={index === props.selectedIndex}
          value={date.valueOf()}
        >
          {date.toDateString()}
        </option>
      );
    })}
  </Form.Control>
);
