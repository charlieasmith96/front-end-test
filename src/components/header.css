.nav-bar {
    display: flex;
    height: 100px;
    background-color: #12A5ED;
}

.nav-bar-text {
    font-size: 70px;
    color: white;
    margin-left: 20px;
    margin-top: 0px
}

.nav-bar-logout {
    position: absolute;
    font-size: 35px;
    color: white;
    margin-top: 12px;
    right: 20px;
    padding: 10px;
    background: none;
	border: none;
	padding: 0;
	outline: none;
}