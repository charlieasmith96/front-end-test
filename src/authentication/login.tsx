import React, { useState, SyntheticEvent, useContext } from 'react'
import { AuthContextProps, AuthContext } from './auth-context-provider'
import { Button, FormGroup, FormControl } from "react-bootstrap";
import { Error } from '../components/error';
import './login.css'

interface LoginProps extends AuthContextProps {}

export const Login = (props: LoginProps) => {

    const [identifiant, setIdentifiant] = useState("");
    const [password, setPassword] = useState("");
  
    function validateForm() {
      return identifiant.length > 0 && password.length > 0;
    }
  
    function handleSubmit(event: SyntheticEvent) {
      event.preventDefault();
      props.login({identifiant, password})
    }

    return (
      <div className="background">
    <div className="login-container">
      <h1 className="login-title">Welcome, please log in</h1>
      <form onSubmit={handleSubmit}>
        <div className={"login"} >
        <FormGroup controlId="email">
          <h2 className="field-label"> Username</h2>
          <FormControl
            data-testid="test-username-id"
            autoFocus
            className="form-field"
            type="text"
            value={identifiant}
            onChange={e => setIdentifiant(e.target.value)}
          />
        </FormGroup>
        <FormGroup className={"login"}controlId="password">
          <h2 className="field-label">Password</h2>
          <FormControl
            data-testid="test-password-id"
            value={password}
            className="form-field"
            onChange={e => setPassword(e.target.value)}
            type="password"
          />
        </FormGroup>
        <Button className="login-button" data-testid='test-button' block disabled={!validateForm()} type="submit">
          Log in
        </Button>
      </div>
      </form>
      {props.authenticationError && <Error errorMessage={props.authenticationError}/>}

    </div>
    </div>
    )
 }