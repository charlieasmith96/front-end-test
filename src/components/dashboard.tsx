import React, { useState, useEffect, useContext, SyntheticEvent } from 'react'
import { getBandwidthDataByDateWithSessionToken, BandwidthRequest, AGGREGATE } from '../api/bandwidth-api'
import { RootContext } from '../authentication/auth-context-provider';
import { Chart, BroadcastData } from './chart';
import { generateDates, subtractDate} from '../helpers/date-utils'
import {Error} from './error';
import { DropdownSelect} from './dropdown-select'
import { Header } from '../components/header'
import { LegendItem } from './legend-item'
import { processBandwidthData, convertBitValueToGigabitValue } from '../helpers/bandwidth-data-processor';

import './dashboard.css'
import { UNEXPECTED_ERROR_STRING, UNAUTHORISED_ERROR_STRING } from './error-string-constants';

export enum DATA_TYPES { CDN = 'cdn', P2P = 'p2p'}

export const Dashboard = () => {

    const { getAuthenticationBody } = useContext(RootContext);

    const [cdnData, setCdnData] = useState<BroadcastData>();
    const [p2pData, setP2pData] = useState<BroadcastData>();
    const [startDate, setStartDate] = useState(subtractDate(5))
    const [endDate, setEndDate] = useState(Date.now())
    const [dates, setDates] = useState<Date[]>([])
    const [cdnMax, setCdnMax] = useState<number>();
    const [p2pAndCdnMax, setP2pAndCdnMax] = useState<number>();
    const [errorMessage, setErrorMessage] = useState<string | null>(null);


     useEffect(() => {
        var bandwidthRequest = 
        {session_token: getAuthenticationBody(),
             from: startDate,
             to: endDate
            }

        getBandwidthDataByDateWithSessionToken(bandwidthRequest)
        .then((response) => { 
            let processedData = processBandwidthData(response);
            
            setCdnData(processedData[0])
            setP2pData(processedData[1])
        }).catch((err : any) => { 
            if (err.status === 403) {
                setErrorMessage(UNAUTHORISED_ERROR_STRING);
            } else {
                setErrorMessage(UNEXPECTED_ERROR_STRING);
            }
        })

        var bandwidthRequestForMax : BandwidthRequest = {session_token: getAuthenticationBody(),
            from: startDate,
            to: endDate,
            aggregate: AGGREGATE.MAX
           }

        getBandwidthDataByDateWithSessionToken(bandwidthRequestForMax)
        .then((response) => {
            setCdnMax(convertBitValueToGigabitValue(response[DATA_TYPES.CDN]));
            setP2pAndCdnMax(convertBitValueToGigabitValue(response[DATA_TYPES.P2P] + response[DATA_TYPES.CDN]));
        }).catch((err : any) => { 
            if (err.status === 403) {
                setErrorMessage(UNAUTHORISED_ERROR_STRING);
            } else {
                setErrorMessage(UNEXPECTED_ERROR_STRING);
            }
        })
        setDates(generateDates())

     }, [startDate, endDate])

    const handleStartDateChange = (event : any) => {
        setStartDate(event.target.value)
    }

    const handleEndDateChange = (event : any) => {
        setEndDate(event.target.value)
    }

    return (
        <div>
            <Header/>
            {!errorMessage ? <div>
            <div className='chart-container'>
                <h2>Bandwidth</h2>
               <LegendItem label="P2P" colour="blue"/>
               <LegendItem label="CDN" colour="pink"/>
               <LegendItem label={`Maximum CDN + P2P: ${p2pAndCdnMax} Gbp/s`} colour="green"/>
               <LegendItem label={`Maximum CDN: ${cdnMax} Gbp/s`} colour="red"/>
                <p className="axis-label">Gbp/s</p>
                <Chart dataSet1={cdnData} dataSet2={p2pData} 
                endDate={endDate} startDate={startDate} p2pAndCdnMax={p2pAndCdnMax} cdnMax={cdnMax}/>
                <div className={"dropdown-container"}>
                    <DropdownSelect dates={dates} onChangeHandler={handleStartDateChange} selectedIndex={5}/>
                    <DropdownSelect dates={dates} onChangeHandler={handleEndDateChange} selectedIndex={0}/>
                </div>
            </div>
            </div>: (<Error className="dashboard-error" errorMessage={errorMessage}/>)}
        </div>
    )
}