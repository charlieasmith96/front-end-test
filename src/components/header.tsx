import React, { useContext } from 'react';
import './header.css';
import { AuthContext } from '../authentication/auth-context-provider';

export const Header = () => {
  const { logout, isAuthenticated } = useContext(AuthContext);

  return (
    <div className="nav-bar">
      <div className="page-container">
        <h1 className="nav-bar-text">Data Dashboard</h1>
        {isAuthenticated && 
            <button data-testid="logout-button" className="nav-bar-logout" onClick={() => logout()}>Log out</button>}
      </div>
    </div>
  );
};
