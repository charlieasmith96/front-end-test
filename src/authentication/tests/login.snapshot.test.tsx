import React from 'react';
import renderer from 'react-test-renderer'
import { Header } from '../../components/header';
import { Login } from '../login';

const testRootContext = {
    isAuthenticated: true,
    login: jest.fn(),
    getAuthenticationBody: jest.fn(),
    logout: jest.fn(),
    authenticationError: null
  }

  const testRootContextWithError = {
      ...testRootContext,
      authenticationError: "test-error"
  }

it('should render login page correctly', () => {
    const tree = renderer.create(<Login {...testRootContext}/>)
    expect(tree).toMatchSnapshot();
})

it('should render login page correctly with error', () => {
    const tree = renderer.create(<Login {...testRootContextWithError}/>)
    expect(tree).toMatchSnapshot();
})