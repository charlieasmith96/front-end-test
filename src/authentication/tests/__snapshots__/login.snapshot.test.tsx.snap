// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render login page correctly 1`] = `
<div
  className="background"
>
  <div
    className="login-container"
  >
    <h1
      className="login-title"
    >
      Welcome, please log in
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="login"
      >
        <div
          className="form-group"
        >
          <h2
            className="field-label"
          >
             Username
          </h2>
          <input
            autoFocus={true}
            className="form-field form-control"
            data-testid="test-username-id"
            id="email"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="login form-group"
        >
          <h2
            className="field-label"
          >
            Password
          </h2>
          <input
            className="form-field form-control"
            data-testid="test-password-id"
            id="password"
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <button
          className="login-button btn btn-primary btn-block"
          data-testid="test-button"
          disabled={true}
          type="submit"
        >
          Log in
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`should render login page correctly with error 1`] = `
<div
  className="background"
>
  <div
    className="login-container"
  >
    <h1
      className="login-title"
    >
      Welcome, please log in
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="login"
      >
        <div
          className="form-group"
        >
          <h2
            className="field-label"
          >
             Username
          </h2>
          <input
            autoFocus={true}
            className="form-field form-control"
            data-testid="test-username-id"
            id="email"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="login form-group"
        >
          <h2
            className="field-label"
          >
            Password
          </h2>
          <input
            className="form-field form-control"
            data-testid="test-password-id"
            id="password"
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <button
          className="login-button btn btn-primary btn-block"
          data-testid="test-button"
          disabled={true}
          type="submit"
        >
          Log in
        </button>
      </div>
    </form>
    <div
      className="error-message undefined"
    >
      test-error
    </div>
  </div>
</div>
`;
